overlay/master.cpp

```diff
@@ -1551,6 +1563,7 @@ class ManagerProcess : public ProtobufProcess<ManagerProcess>
     CHECK_NOTNULL(replicatedLog.get());
 
     recovering = true;
+    metrics->recovering = 1;
```

Comment:
Master module has three states: stand by, recovering, and recovered. With `recovering` metric, how can we differentiate the first and the last states?

When module is restarted by supervisor module, should we reset `recovering` metric explicitly?

Reply:
> Master module has three states: stand by, recovering, and recovered. With recovering metric, how can we differentiate the first and the last states?

Hmm, probably I am missing something. My idea is to alert if any replica stays in `recovering` state for too long. By `stand by` mode, do you mean that there is at most active leader, and all other replicas are standing by? If so, this is the part I do not get. I can map the states you listed to certain implementation details of replicated log, which are not exposed to the outside world. Instead, there is an API of different kind, which is used by the overlay master processes. The best I can think of is, introducing `active` metric, which is set to 0 in every master processes upon start-up, which stays that way if there are no agent registration messages received. Once such a message gets received, the recovery process begins, and `recovering` changes accordingly. Once recovery is completed, `active` is set to 1. Are you thinking of something along those lines? If no, could you please elaborate?

> When module is restarted by supervisor module, should we reset recovering metric explicitly?

I don't think so, because the metric gets recreated upon restart anyways.

Reply:
It looks like we call `recover()` function when we receive the first `RegisterAgentMessage` which is sent by agent module to that master node when it detects that the master node is elected as a new leader. When new election happens, we do noting in master module. It means that we only can set `active` to 1, but never can set it back to `0`.

I think we are fine if you add a counter for `RegisterAgentMessage`s: if the metric is 0 and `recovering` is 0 then we never tried to recover module's state from the replicated log; if the metric is 1 or bigger and `recovering` is 1 then the module is recovered.